relayer/spy_relayer/src/backends/default/listener.ts

```diff
@@ -0,0 +1,272 @@
+      return false;
+    }
+    return true;
+  }
+
+  /** Verify this payload has a fee specified for relaying. */
+  verifyFeeSpecified(payload: ParsedTransferPayload): boolean {
+    /**
+     * TODO: simulate gas fees / get notional from coingecko and ensure the fees cover the relay.
+     *       We might just keep this check here but verify the notional is enough to pay the gas
+     *       fees in the actual relayer. That way we can retry up to the max number of retries
+     *       and if the gas fluctuates we might be able to make it still.
+     */
+
+    /** Is the specified fee sufficient to relay? */
+    const sufficientFee = payload.fee && payload.fee > BigInt(0);
+
+    if (!sufficientFee) {
+      this.logger.debug("Token transfer does not have a sufficient fee.");
+      return false;
+    }
+    return true;
+  }
+
+  /** Verify the the token in this payload in the approved token list. */
+  verifyIsApprovedToken(payload: ParsedTransferPayload): boolean {
+    const env = getListenerEnvironment();
+    let originAddressNative: string;
+    try {
+      originAddressNative = tryHexToNativeString(
+        payload.originAddress,
+        payload.originChain
+      );
+    } catch (e: any) {
+      return false;
+    }
+
+    // Token is in the SUPPORTED_TOKENS env var config
+    const isApprovedToken = env.supportedTokens.find((token) => {
+      return (
+        originAddressNative &&
+        token.address.toLowerCase() === originAddressNative.toLowerCase() &&
+        token.chainId === payload.originChain
+      );
+    });
+
+    if (!isApprovedToken) {
+      this.logger.debug("Token transfer is not for an approved token.");
+      return false;
+    }
+
+    return true;
+  }
+
+  /** Parses a raw VAA byte array
+   *
+   * @throws when unable to parse the VAA
+   */
+  public async parseVaa(rawVaa: Uint8Array): Promise<ParsedVaa<Uint8Array>> {
+    let parsedVaa: ParsedVaa<Uint8Array> | null = null;
+
+    try {
+      parsedVaa = await parseVaaTyped(rawVaa);
+    } catch (e) {
+      this.logger.error("Encountered error while parsing raw VAA " + e);
+    }
+    if (!parsedVaa) {
+      throw new Error("Unable to parse the specified VAA.");
+    }
+
+    return parsedVaa;
+  }
+
+  /** Parse the VAA and return the payload nicely typed */
+  public async parsePayload(
+    rawPayload: Uint8Array
+  ): Promise<ParsedTransferPayload> {
+    let parsedPayload: any;
+    try {
+      parsedPayload = parseTransferPayload(Buffer.from(rawPayload));
+    } catch (e) {
+      this.logger.error("Encountered error while parsing vaa payload" + e);
+    }
+
+    if (!parsedPayload) {
+      this.logger.debug("Failed to parse the transfer payload.");
+      throw new Error("Could not parse the transfer payload.");
+    }
+    return parsedPayload;
+  }
+
+  /** Verify this is a VAA we want to relay. */
+  public async validate(
+    rawVaa: Uint8Array
+  ): Promise<ParsedVaa<ParsedTransferPayload> | string> {
```

Comment:
I'm not actually sure which is the best approach to write idiomatic typescript here:

* Just throwing `Error()` on any problem
* The return being the correct type OR a string